### PR TITLE
Implementation of new Smart Solr 2 design

### PR DESCRIPTION
**Pull request by noblepaul**
design can be found [here](https://docs.google.com/document/d/1xdxpzUNmTZbk0vTMZqfen9R3ArdHokLITdiISBxCFUg/edit?usp=sharing) 

There are a bunch of `nocommit` added in the code. These are mostly debugging messages. they have to be removed right before we merge this

What's left?

* command to upgrade
* command to downgrade

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java
Comment (donjaime) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:452]: Should nodeName be a constructor param?

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java
Comment (donjaime) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java]: nit: But what is Solr clouds code style regarding conditionals and curlies being optional? Granted in this instance all of these are single return statements. So not a huge issue. But multi-statement conditionals without curlies has bitten so many people before that we enforce them in our Java code style at FullStory.

File: solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java
Comment (donjaime) [inline solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java:242]: Delete commented out changes?

File: solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java
Comment (donjaime) [inline solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java:270]: Does this need a `&& !cmd.ops.get().isEmpty()`  to match the pre ops?

File: solr/core/src/java/org/apache/solr/cloud/overseer/ZkWriteCommand.java
Comment (donjaime) [inline solr/core/src/java/org/apache/solr/cloud/overseer/ZkWriteCommand.java:26]: What is the javadoc above documenting?

File: solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java
Comment (donjaime) [inline solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java]: Should this be a debug log?

File: solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java
Comment (donjaime) [inline solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java]: Should this be a debug log?

File: solr/core/src/java/org/apache/solr/cloud/overseer/ClusterStateMutator.java
Comment (donjaime) [inline solr/core/src/java/org/apache/solr/cloud/overseer/ClusterStateMutator.java]: Should this still be a debug log? Why elevate to info?

File: solr/core/src/test/org/apache/solr/cloud/CollectionsAPISolrJTest.java
Comment (donjaime) [inline solr/core/src/test/org/apache/solr/cloud/CollectionsAPISolrJTest.java:594]: Do we run these tests twice? Once with `USE_PER_REPLICA_STATE` `false` (as it the default) and one with it set to `true`?

I'm not seeing it set to `true` anywhere. Are you basically manually changing your environment to run these tests?

**Review by donjaime — COMMENTED** [review]
Here is an initial (superficial) review pass.

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:452]: We don't need it actually. It was added for debugging and will be removed before I merge this PR. I've added a `nocommit` message to the field

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java]: the coding style allows no curlies for simple `if` checks

File: solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java:242]: These were left for anyone who wishes to debug something. I have added a `nocommit` will be removed right before merging the PR

File: solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java:270]: well the `isEmpty()` check can be removed in the pre-ops also as the ` PerReplicaStates.presist()` checks them anyway

File: solr/core/src/java/org/apache/solr/cloud/overseer/ZkWriteCommand.java
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ZkWriteCommand.java:26]: There was a field there. this is a leftover. removed

File: solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java:1433]: true

File: solr/core/src/test/org/apache/solr/cloud/CollectionsAPISolrJTest.java
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/cloud/CollectionsAPISolrJTest.java:594]: Yes, We run these tests with that property `true` and `false`

We couldn't find a better way to do this yet

Comment (donjaime) [reply solr/core/src/test/org/apache/solr/cloud/CollectionsAPISolrJTest.java:594]: Building it this way is fine. But for clarity what I meant was that the default value is false. And these tests will *not* be run by any CI system. Neither ours, nor Solrs. Unless we setup configuration with the value set to `true`. But then you won't be testing the code paths with the value set to `false` anymore.

I think of testing as a single suite testing all permutations. Not a runtime-sensitive thing. If I were settings this up with continuous tests, I would want to run these parameterized tests for each value of `USE_PER_REPLICA_STATE`. Once for `false` and once for `true`.

File: solr/core/src/test/org/apache/solr/handler/TestStressThreadBackup.java
Comment (donjaime) [inline solr/core/src/test/org/apache/solr/handler/TestStressThreadBackup.java:64]: Intentional?

File: solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java
Comment (donjaime) [inline solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java]: Does this method need to be synchronized?
This is not a simple accessor. It mutates local state on read. That seems dangerous?

Comment (donjaime) [reply solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java]: It's documented as `for internal use only`. But this is a `public` method. Who are the users and when should this be used?

File: solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java
Comment (donjaime) [inline solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java]: These two `persist` functions have the same public API signature (almost) and no javadoc explaining the difference.

Please add some to avoid confusion about what is doing what.

File: solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java
Comment (donjaime) [inline solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java]: For a second I thought this function was infinitely recursive. But then I see another function `persist` below with nearly identical method signatures. But no javadoc explaining the difference.

File: solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java
Comment (donjaime) [inline solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java]: Same comment. If you are going to shadow function names in public API. Please javadoc.

File: solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java
Comment (donjaime) [inline solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java]: Minor. But weird spacing above. You might want to update your Java code formatting rules and configure Intellij to format on save?

File: solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java
Comment (donjaime) [inline solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java]: Will these ops apply atomically? If they apply serially we might have an issue. Since you queue the ADD before the DELETE I think this means a single replica can have two entries for a brief moment. Where it is both the leader and not the leader.

Maybe you should delete and add in that order. And know that you can occasionally encounter situations where a replica might (for a brief moment) not have any entry in the per-replica states list.

If the ZkClient multi operation stuff applies it atomically, then I think we are fine.

File: solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java
Comment (donjaime) [inline solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java]: Same comment here about atomic updates and which partial state is preferable. Either paradoxical entries, or temporarily have no entry (delete first) while you queue the add to update it.

File: solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java
Comment (donjaime) [inline solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java]: What does this do?

**Review by donjaime — COMMENTED** [review]
Some more feedback.

Overall the approach seems fine. Mostly localized questions. And general superficial notes about missing JavaDoc and code style issues.

File: solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java
Comment (noblepaul) [reply solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java]: * ZK `multi()` is atomic
* The class is written in such a way that, if there are multiple entries for the same replica, only the newest one is picked

File: solr/core/src/test/org/apache/solr/handler/TestStressThreadBackup.java
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/handler/TestStressThreadBackup.java:64]: unintentional, removing

File: solr/core/src/java/org/apache/solr/cloud/ShardLeaderElectionContextBase.java
Comment (EmilyBragg) [inline solr/core/src/java/org/apache/solr/cloud/ShardLeaderElectionContextBase.java:187]: This isn't a function that was written in this PR, but asking in case you know - the `getStateFormat()` call appears to be an equality check (included in comment at bottom), is there a particular reason it's as generically named as it is? Mainly just asking because I can't tell if `getStateFormat()` indicates some subtlety beyond "the znodes match" even after digging in a bit

```
 public int getStateFormat() {
    return ZkStateReader.CLUSTER_STATE.equals(znode) ? 1 : 2;
 }
```

Comment (EmilyBragg) [reply solr/core/src/java/org/apache/solr/cloud/ShardLeaderElectionContextBase.java:187]: (Just want to make sure if there is a particular subtle aspect there that I'm keeping it in mind while reviewing :) )

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java
Comment (EmilyBragg) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java]: nit: whitespace and spelling

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java
Comment (EmilyBragg) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:1618]: Unlikely bc this is a bad/invalid state, or how?

File: solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java
Comment (EmilyBragg) [inline solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java:126]: Do we want to log this somehow if it's removing things? That sounds like it could be a subtle bug to find later if it somehow did encounter this.

File: solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java
Comment (EmilyBragg) [inline solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java:142]: ought we break after this gets set?

File: solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ZkStateWriter.java:126]: It's not removing , it's just taking them out and executing right away

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1618]: yeah, it has to be a bad/invalid state

File: solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java
Comment (noblepaul) [reply solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java]: it's possible that there are multiple entries for same replica as follows


- `core_node_1:12:A`
- `core_node_1:13:D`

Essentially there are multiple entries for the same replica. In this case `13` is the latest and `12` is a duplicate.  It's only used in the `toString()` method

File: solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java
Comment (noblepaul) [reply solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java]: This method must go. It doesn't belong here

**Review by donjaime — APPROVED** [review]
Approving so that you can proceed with more rigorous load testing in GCP.
Though I wouldn't mind a spot check from @claygoddard or @dragonsinth on the overall approach. I think that can happen as followups.

LGTM

**Comment by claygoddard** [comment]
> Approving so that you can proceed with more rigorous load testing in GCP.
> Though I wouldn't mind a spot check from @claygoddard or @dragonsinth on the overall approach. I think that can happen as followups.
> 
> LGTM

It's been a little hard to make time this week to give this a good look. I will try to make some time though.

**Pull request closed without merging**
